### PR TITLE
OSC/UCX: fixed zero-size window processing

### PR DESCRIPTION
**Pull request by hoopoepg**
- added processing of zero-size MPI window

fixes https://github.com/open-mpi/ompi/issues/5873

Signed-off-by: Sergey Oblomov <sergeyo@mellanox.com>

### CODE REVIEW

File: ompi/mca/osc/ucx/osc_ucx_comm.c [hunk 1]
```diff
@@ -380,6 +380,13 @@ int ompi_osc_ucx_put(const void *origin_addr, int origin_count, struct ompi_data
         }
     }
 
+    if (!((module->win_info_array[target]).rkey_init) && (target_count > 0)) {
```
Comment (yosefe) [inline ompi/mca/osc/ucx/osc_ucx_comm.c]: maybe add this as macro/common function?
also looks like two pairs of redundant () around !((module->win_info_array[target]).rkey_init)

File: ompi/mca/osc/ucx/osc_ucx_comm.c [hunk 2]
```diff
@@ -380,6 +380,13 @@ int ompi_osc_ucx_put(const void *origin_addr, int origin_count, struct ompi_data
...
+        OSC_UCX_VERBOSE(1, "target window has no rkey or has zero size");
```
Comment (yosefe) [inline ompi/mca/osc/ucx/osc_ucx_comm.c]: "window with non-zero length does not have an rkey"

File: ompi/mca/osc/ucx/osc_ucx_component.c [hunk 3]
```diff
@@ -521,22 +526,26 @@ static int component_select(struct ompi_win_t *win, void **base, size_t size, in
...
+        memcpy_off(my_info, base, sizeof(uint64_t), info_offset);
     } else {
-        memcpy(my_info, &zero, sizeof(uint64_t));
+        memcpy_off(my_info, &zero, sizeof(uint64_t), info_offset);
     }
-    memcpy((void *)((char *)my_info + sizeof(uint64_t)), &state_base, sizeof(uint64_t));
-    memcpy((void *)((char *)my_info + 2 * sizeof(uint64_t)), rkey_buffer, rkey_buffer_size);
-    memcpy((void *)((char *)my_info + 2 * sizeof(uint64_t) + rkey_buffer_size),
-           state_rkey_buffer, state_rkey_buffer_size);
+    memcpy_off(my_info, &state_base, sizeof(uint64_t), info_offset);
```
Comment (yosefe) [inline ompi/mca/osc/ucx/osc_ucx_component.c]: maybe add the window size as another uint64_t in the beginning (in addition to state_base and rkey_buffer_size)?

File: ompi/mca/osc/ucx/osc_ucx_comm.c (hunk 1 above)
Comment (hoopoepg) [reply ompi/mca/osc/ucx/osc_ucx_comm.c]: added macro

File: ompi/mca/osc/ucx/osc_ucx_comm.c (hunk 2 above)
Comment (hoopoepg) [reply ompi/mca/osc/ucx/osc_ucx_comm.c]: fixed

File: ompi/mca/osc/ucx/osc_ucx_component.c (hunk 3 above)
Comment (hoopoepg) [reply ompi/mca/osc/ucx/osc_ucx_component.c]: didn't get your comment, size is added on next line (line 544 of updated file)

Comment (hoopoepg) [reply ompi/mca/osc/ucx/osc_ucx_component.c]: hmmmm, size is size_t type, not uint64_t, merge into 3 * sizeof(uint64_t) ?

**Comment by hoopoepg** [comment]
@yosefe ok to approve or more comments?

File: ompi/mca/osc/ucx/osc_ucx_component.c (hunk 3 above)
Comment (hoopoepg) [reply ompi/mca/osc/ucx/osc_ucx_component.c]: updated

**Comment by hoopoepg** [comment]
@yosefe I will squash commits after tests are completed, will ping you when it ready to commit

**Comment by hoopoepg** [comment]
bot:retest

**Comment by hoopoepg** [comment]
@yosefe ok to merge?

**Comment by yosefe** [comment]
@hoopoepg need for v4.0.x

**Pull request merged**